### PR TITLE
(WIP) Pointnet

### PR DESCRIPTION
**Pull request by tristanheywood**

### CODE REVIEW

File: models/PointNet/modules.py [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+    def __init__(self, local_nn, global_nn):
+        super().__init__()
+
+        self.local_nn = MLP(local_nn)
+        self.global_nn = MLP(global_nn)
+
+    def forward(self, x, batch):
+        
+        x = self.local_nn(x)
+        x = global_max_pool(x, batch)
```
Comment (tchaton) [inline models/PointNet/modules.py]: global_max_pooling uses scatter_max in the backend. If the data are dense [B, P, F], use maxpooling, else scatter_max -> gain in speed.

**Comment by tchaton** [comment]
Good and clean PR @tristanheywood. Please, add a dim check and use global_max accordingly for speed up.

Best,
Thomas Chaton.

**Comment by tchaton** [comment]
Hey @tristanheywood,

I have added an example to switch between aggregation. However, it doesn't seem to change much the speed between both implementation.

Best,
Thomas Chaton.

**Comment by tchaton** [comment]
```
Model size = 3536213
Access tensorboard with the following command <tensorboard --logdir=/home/thomas/HELIX/research/deeppointcloud-benchmarks/outputs/2020-01-03/19-24-22/tensorboard>
EPOCH 1 / 350
100%|██████████████████████████████████████████████████████████████████████████▊| 587/588 [02:15<00:00,  4.32it/s, data_loading=0.002, iteration=0.255, train_acc=76.70, train_macc=72.04, train_miou=58.29]
Traceback (most recent call last):
  File "train.py", line 125, in <module>
    main()
  File "/home/thomas/.cache/pypoetry/virtualenvs/superpoint-graph-job-py3.6/lib/python3.6/site-packages/hydra/main.py", line 24, in decorated_main
    strict=strict,
  File "/home/thomas/.cache/pypoetry/virtualenvs/superpoint-graph-job-py3.6/lib/python3.6/site-packages/hydra/_internal/utils.py", line 174, in run_hydra
    overrides=args.overrides,
  File "/home/thomas/.cache/pypoetry/virtualenvs/superpoint-graph-job-py3.6/lib/python3.6/site-packages/hydra/_internal/hydra.py", line 86, in run
    job_subdir_key=None,
  File "/home/thomas/.cache/pypoetry/virtualenvs/superpoint-graph-job-py3.6/lib/python3.6/site-packages/hydra/plugins/common/utils.py", line 109, in run_job
    ret.return_value = task_function(task_cfg)
  File "train.py", line 121, in main
    run(cfg, model, dataset, device, tracker, checkpoint)
  File "train.py", line 71, in run
    train(epoch, model, train_loader, device, tracker, checkpoint)
  File "train.py", line 36, in train
    model.optimize_parameters()
  File "/home/thomas/HELIX/research/deeppointcloud-benchmarks/models/base_model.py", line 68, in optimize_parameters
    self.forward()               # first call forward to calculate intermediate results
  File "/home/thomas/HELIX/research/deeppointcloud-benchmarks/models/PointNet/nn.py", line 35, in forward
    x = self.pointnet_seg(self.input.pos, self.input.batch)
  File "/home/thomas/.cache/pypoetry/virtualenvs/superpoint-graph-job-py3.6/lib/python3.6/site-packages/torch/nn/modules/module.py", line 493, in __call__
    result = self.forward(*input, **kwargs)
  File "/home/thomas/HELIX/research/deeppointcloud-benchmarks/models/PointNet/modules.py", line 96, in forward
    x = self.input_stn(x, batch)
  File "/home/thomas/.cache/pypoetry/virtualenvs/superpoint-graph-job-py3.6/lib/python3.6/site-packages/torch/nn/modules/module.py", line 493, in __call__
    result = self.forward(*input, **kwargs)
  File "/home/thomas/HELIX/research/deeppointcloud-benchmarks/models/PointNet/modules.py", line 38, in forward
    return super().forward(x, x, batch)
  File "/home/thomas/HELIX/research/deeppointcloud-benchmarks/models/core_transforms.py", line 35, in forward
    global_feature = self.nn(feat_x, batch)
  File "/home/thomas/.cache/pypoetry/virtualenvs/superpoint-graph-job-py3.6/lib/python3.6/site-packages/torch/nn/modules/module.py", line 493, in __call__
    result = self.forward(*input, **kwargs)
  File "/home/thomas/HELIX/research/deeppointcloud-benchmarks/models/PointNet/modules.py", line 22, in forward
    x = self.global_nn(x)
  File "/home/thomas/.cache/pypoetry/virtualenvs/superpoint-graph-job-py3.6/lib/python3.6/site-packages/torch/nn/modules/module.py", line 493, in __call__
    result = self.forward(*input, **kwargs)
  File "/home/thomas/.cache/pypoetry/virtualenvs/superpoint-graph-job-py3.6/lib/python3.6/site-packages/torch/nn/modules/container.py", line 92, in forward
    input = module(input)
  File "/home/thomas/.cache/pypoetry/virtualenvs/superpoint-graph-job-py3.6/lib/python3.6/site-packages/torch/nn/modules/module.py", line 493, in __call__
    result = self.forward(*input, **kwargs)
  File "/home/thomas/.cache/pypoetry/virtualenvs/superpoint-graph-job-py3.6/lib/python3.6/site-packages/torch/nn/modules/container.py", line 92, in forward
    input = module(input)
  File "/home/thomas/.cache/pypoetry/virtualenvs/superpoint-graph-job-py3.6/lib/python3.6/site-packages/torch/nn/modules/module.py", line 493, in __call__
    result = self.forward(*input, **kwargs)
  File "/home/thomas/.cache/pypoetry/virtualenvs/superpoint-graph-job-py3.6/lib/python3.6/site-packages/torch/nn/modules/batchnorm.py", line 83, in forward
    exponential_average_factor, self.eps)
  File "/home/thomas/.cache/pypoetry/virtualenvs/superpoint-graph-job-py3.6/lib/python3.6/site-packages/torch/nn/functional.py", line 1693, in batch_norm
    raise ValueError('Expected more than 1 value per channel when training, got input size {}'.format(size))
ValueError: Expected more than 1 value per channel when training, got input size torch.Size([1, 512])
```

**Comment by tristanheywood** [comment]
@tchaton when I was training with shapenet each sample in the batch had a different number of points. I was wondering why that was - I thought shapenet samples always have 2048 points.

**Comment by tchaton** [comment]
Hey @tristanheywood, we added transform = T.FixedPoints(dataset_opt.num_points) in the class ShapeNetDataset.

**Comment by tristanheywood** [comment]
oh I see, so PYG's shapenet doesn't automatically give you a fixed number of points - you have to do that with a transform. 

I also use the batch index in the stn to do the matrix multiplication. I think that could be made faster if the samples had a fixed number of points. @tchaton I put the stn in core modules, how can I check if the samples are the same size? Maybe we need to abstract away all the operations that are different for dense vs sparse. We could make a library of operation like max pooling, matrix multiplication etc and in that library check for dense vs sparse.

**Pull request merged**